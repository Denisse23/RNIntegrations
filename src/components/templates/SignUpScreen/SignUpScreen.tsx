import React, { useEffect, useState } from 'react';
import { Button, CustomInput, CustomText } from '@components/atoms';
import { FormLayout } from '@components/organisms';
import { NavigationType } from '@navigation/types';
import { useNavigation } from '@react-navigation/native';
import { Divider, LargeDivider } from '@styles/SharedStyles';
import { INPUT_COMMON_PROPS, TEST_IDS } from '@constants';
import strings from '@localization';
import { StyledContainer } from './SignUpScreen.styles';

const SignUpScreen = () => {
  const navigation = useNavigation<NavigationType>();

  const [emailValue, setEmailValue] = useState('');
  const [firstNameValue, setFirstNameValue] = useState('');
  const [lastNameValue, setLastNameValue] = useState('');
  const [passwordValue, setPasswordValue] = useState('');
  const [confirmPasswordValue, setConfirmPasswordValue] = useState('');
  const [emailError, setEmailError] = useState<string | null>(null);
  const [firstNameError, setFirstNameError] = useState<string | null>(null);
  const [lastNameError, setLastNameError] = useState<string | null>(null);
  const [passwordError, setPasswordError] = useState<string | null>(null);
  const [confirmPasswordError, setConfirmPasswordError] = useState<
    string | null
  >(null);

  useEffect(() => {
    if (confirmPasswordValue && !passwordError) {
      if (confirmPasswordValue !== passwordValue) {
        setConfirmPasswordError(strings.authentication.confirmPasswordError);
      } else {
        setConfirmPasswordError(null);
      }
    }
  }, [confirmPasswordValue, passwordError, passwordValue]);

  const signUpDisabled = () => {
    return (
      !emailValue ||
      !firstNameValue ||
      !lastNameValue ||
      !passwordValue ||
      !confirmPasswordValue ||
      !!emailError ||
      !!firstNameError ||
      !!lastNameError ||
      !!passwordError ||
      !!confirmPasswordError
    );
  };

  return (
    <FormLayout testID={TEST_IDS.templates.signUpScreen}>
      <StyledContainer>
        <CustomText size="xxl" type="bold">
          {strings.authentication.createAnAccount}
        </CustomText>
        <LargeDivider />
        <CustomInput
          {...INPUT_COMMON_PROPS.email}
          label={strings.authentication.email}
          value={emailValue}
          onChangeText={setEmailValue}
          error={emailError}
          onChangeError={setEmailError}
        />
        <Divider />
        <CustomInput
          {...INPUT_COMMON_PROPS.name}
          label={strings.authentication.firstName}
          value={firstNameValue}
          onChangeText={setFirstNameValue}
          error={firstNameError}
          onChangeError={setFirstNameError}
        />
        <Divider />
        <CustomInput
          {...INPUT_COMMON_PROPS.name}
          label={strings.authentication.lastName}
          value={lastNameValue}
          onChangeText={setLastNameValue}
          error={lastNameError}
          onChangeError={setLastNameError}
        />
        <Divider />
        <CustomInput
          {...INPUT_COMMON_PROPS.newPassword}
          label={strings.authentication.password}
          value={passwordValue}
          onChangeText={setPasswordValue}
          error={passwordError}
          onChangeError={setPasswordError}
        />
        <Divider />
        <CustomInput
          {...INPUT_COMMON_PROPS.password}
          label={strings.authentication.confirmPassword}
          value={confirmPasswordValue}
          onChangeText={setConfirmPasswordValue}
          error={confirmPasswordError}
          onChangeError={setConfirmPasswordError}
        />
        <LargeDivider />
        <LargeDivider />
        <Button
          text={strings.authentication.signUp}
          type="primary"
          disabled={signUpDisabled()}
        />
        <Button
          text={strings.authentication.alreadyHaveAnAccount}
          type="secondaryText"
          onPress={() => navigation.navigate('LoginScreen')}
        />
      </StyledContainer>
    </FormLayout>
  );
};

export default SignUpScreen;
