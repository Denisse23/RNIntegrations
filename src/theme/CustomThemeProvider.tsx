import React from 'react';
import { useColorScheme } from 'react-native';
import { ThemeProvider } from '@emotion/react';
import { ColorScheme, DarkTheme, LightTheme } from './Themes';

const CustomThemeProvider = ({ children }: { children: React.ReactNode }) => {
  const scheme = useColorScheme();

  return (
    <ThemeProvider theme={scheme === ColorScheme.DARK ? DarkTheme : LightTheme}>
      {children}
    </ThemeProvider>
  );
};

export default CustomThemeProvider;
