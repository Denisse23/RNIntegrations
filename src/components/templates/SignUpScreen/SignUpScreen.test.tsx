import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import CustomThemeProvider from '@theme/CustomThemeProvider';
import renderer from 'react-test-renderer';
import { TEST_IDS } from '@constants';
import SignUpScreen from './SignUpScreen';

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: jest.fn(),
}));

describe('SignUpScreen', () => {
  it('should not enable button if the inputs are empty', async () => {
    const { getByTestId } = render(
      <CustomThemeProvider>
        <SignUpScreen />
      </CustomThemeProvider>,
    );

    const button = getByTestId(TEST_IDS.buttons.signUpButton);

    expect(button.props).toHaveProperty('accessibilityState.disabled', true);
  });

  it('should not enable button if the passwords are different', async () => {
    const { getByTestId } = render(
      <CustomThemeProvider>
        <SignUpScreen />
      </CustomThemeProvider>,
    );

    const emailInput = getByTestId(TEST_IDS.inputs.emailInput);
    const passwordInput = getByTestId(TEST_IDS.inputs.passwordInput);
    const confirmPasswordInput = getByTestId(
      TEST_IDS.inputs.configmPasswordInput,
    );
    await waitFor(() => {
      fireEvent.changeText(emailInput, 'test@gmail.com');
      fireEvent.changeText(passwordInput, 'Test12346!');
      fireEvent.changeText(confirmPasswordInput, 'Test12345!');
    });

    const button = getByTestId(TEST_IDS.buttons.signUpButton);

    expect(button.props).toHaveProperty('accessibilityState.disabled', false);
  });

  it('should not enable button if the password is short', async () => {
    const { getByTestId } = render(
      <CustomThemeProvider>
        <SignUpScreen />
      </CustomThemeProvider>,
    );

    const emailInput = getByTestId(TEST_IDS.inputs.emailInput);
    const passwordInput = getByTestId(TEST_IDS.inputs.passwordInput);
    const confirmPasswordInput = getByTestId(
      TEST_IDS.inputs.configmPasswordInput,
    );
    await waitFor(() => {
      fireEvent.changeText(emailInput, 'test@gmail.com');
      fireEvent.changeText(passwordInput, 'Test1');
      fireEvent.changeText(confirmPasswordInput, 'Test1');
    });

    const button = getByTestId(TEST_IDS.buttons.signUpButton);

    expect(button.props).toHaveProperty('accessibilityState.disabled', false);
  });

  it('should not enable button if the password is missing special character', async () => {
    const { getByTestId } = render(
      <CustomThemeProvider>
        <SignUpScreen />
      </CustomThemeProvider>,
    );

    const emailInput = getByTestId(TEST_IDS.inputs.emailInput);
    const passwordInput = getByTestId(TEST_IDS.inputs.passwordInput);
    const confirmPasswordInput = getByTestId(
      TEST_IDS.inputs.configmPasswordInput,
    );
    await waitFor(() => {
      fireEvent.changeText(emailInput, 'test@gmail.com');
      fireEvent.changeText(passwordInput, 'Test1234');
      fireEvent.changeText(confirmPasswordInput, 'Test1234');
    });

    const button = getByTestId(TEST_IDS.buttons.signUpButton);

    expect(button.props).toHaveProperty('accessibilityState.disabled', false);
  });

  it('should enable button if the inputs are not empty and passwords are the same and strong', async () => {
    const { getByTestId } = render(
      <CustomThemeProvider>
        <SignUpScreen />
      </CustomThemeProvider>,
    );

    const emailInput = getByTestId(TEST_IDS.inputs.emailInput);
    const passwordInput = getByTestId(TEST_IDS.inputs.passwordInput);
    const confirmPasswordInput = getByTestId(
      TEST_IDS.inputs.configmPasswordInput,
    );
    await waitFor(() => {
      fireEvent.changeText(emailInput, 'test@gmail.com');
      fireEvent.changeText(passwordInput, 'Test12345!');
      fireEvent.changeText(confirmPasswordInput, 'Test12345!');
    });

    const button = getByTestId(TEST_IDS.buttons.signUpButton);

    expect(button.props).toHaveProperty('accessibilityState.disabled', false);
  });

  it('should match the snapshot', () => {
    const tree = renderer.create(
      <CustomThemeProvider>
        <SignUpScreen />
      </CustomThemeProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
});
