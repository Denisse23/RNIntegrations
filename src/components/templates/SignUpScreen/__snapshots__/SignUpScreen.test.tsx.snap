// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpScreen should match the snapshot 1`] = `
<RCTSafeAreaView
  accessibilityLabel="sign-up-screen"
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
  testID="sign-up-screen"
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Array [
              Object {
                "padding": 37.5,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "justifyContent": "center",
              },
              Array [
                Object {
                  "marginTop": 117.1875,
                },
                undefined,
              ],
            ]
          }
        >
          <Text
            size="xxl"
            style={
              Array [
                Object {
                  "color": "#353434",
                  "fontSize": 46.875,
                  "fontWeight": "700",
                },
                undefined,
              ]
            }
            type="bold"
          >
            Creat an account
          </Text>
          <View
            style={
              Array [
                Object {
                  "height": 75,
                },
                undefined,
              ]
            }
          />
          <View
            error={null}
            style={
              Array [
                Object {
                  "backgroundColor": "#D5D5D533",
                  "borderBottomWidth": 2.34375,
                  "borderColor": "#D5D5D5",
                  "borderLeftWidth": 7.03125,
                  "borderRadius": 75,
                  "borderRightWidth": 2.34375,
                  "borderTopWidth": 7.03125,
                  "height": 117.1875,
                  "paddingHorizontal": 18.75,
                },
                undefined,
              ]
            }
          >
            <TextInput
              accessibilityLabel="email-input"
              autoCapitalize="none"
              keyboardType="email-address"
              maxLength={100}
              onChangeText={[Function]}
              placeholder="Email"
              placeholderTextColor="#A9A8A3"
              style={
                Array [
                  Object {
                    "color": "#353434",
                    "fontSize": 32.8125,
                    "height": "100%",
                    "paddingHorizontal": 43.9453125,
                    "paddingTop": 0,
                    "position": "absolute",
                    "width": "100%",
                  },
                  undefined,
                ]
              }
              testID="email-input"
              value=""
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "height": 37.5,
                },
                undefined,
              ]
            }
          />
          <View
            error={null}
            style={
              Array [
                Object {
                  "backgroundColor": "#D5D5D533",
                  "borderBottomWidth": 2.34375,
                  "borderColor": "#D5D5D5",
                  "borderLeftWidth": 7.03125,
                  "borderRadius": 75,
                  "borderRightWidth": 2.34375,
                  "borderTopWidth": 7.03125,
                  "height": 117.1875,
                  "paddingHorizontal": 18.75,
                },
                undefined,
              ]
            }
          >
            <TextInput
              maxLength={50}
              onChangeText={[Function]}
              placeholder="First name"
              placeholderTextColor="#A9A8A3"
              style={
                Array [
                  Object {
                    "color": "#353434",
                    "fontSize": 32.8125,
                    "height": "100%",
                    "paddingHorizontal": 43.9453125,
                    "paddingTop": 0,
                    "position": "absolute",
                    "width": "100%",
                  },
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "height": 37.5,
                },
                undefined,
              ]
            }
          />
          <View
            error={null}
            style={
              Array [
                Object {
                  "backgroundColor": "#D5D5D533",
                  "borderBottomWidth": 2.34375,
                  "borderColor": "#D5D5D5",
                  "borderLeftWidth": 7.03125,
                  "borderRadius": 75,
                  "borderRightWidth": 2.34375,
                  "borderTopWidth": 7.03125,
                  "height": 117.1875,
                  "paddingHorizontal": 18.75,
                },
                undefined,
              ]
            }
          >
            <TextInput
              maxLength={50}
              onChangeText={[Function]}
              placeholder="Last name"
              placeholderTextColor="#A9A8A3"
              style={
                Array [
                  Object {
                    "color": "#353434",
                    "fontSize": 32.8125,
                    "height": "100%",
                    "paddingHorizontal": 43.9453125,
                    "paddingTop": 0,
                    "position": "absolute",
                    "width": "100%",
                  },
                  undefined,
                ]
              }
              value=""
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "height": 37.5,
                },
                undefined,
              ]
            }
          />
          <View
            error={null}
            style={
              Array [
                Object {
                  "backgroundColor": "#D5D5D533",
                  "borderBottomWidth": 2.34375,
                  "borderColor": "#D5D5D5",
                  "borderLeftWidth": 7.03125,
                  "borderRadius": 75,
                  "borderRightWidth": 2.34375,
                  "borderTopWidth": 7.03125,
                  "height": 117.1875,
                  "paddingHorizontal": 18.75,
                },
                undefined,
              ]
            }
          >
            <TextInput
              accessibilityLabel="password-input"
              autoCapitalize="none"
              keyboardType="default"
              maxLength={100}
              multiline={false}
              onChangeText={[Function]}
              placeholder="Password"
              placeholderTextColor="#A9A8A3"
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "color": "#353434",
                    "fontSize": 32.8125,
                    "height": "100%",
                    "paddingHorizontal": 43.9453125,
                    "paddingTop": 0,
                    "position": "absolute",
                    "width": "100%",
                  },
                  undefined,
                ]
              }
              testID="password-input"
              textContentType="newPassword"
              value=""
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "height": 37.5,
                },
                undefined,
              ]
            }
          />
          <View
            error={null}
            style={
              Array [
                Object {
                  "backgroundColor": "#D5D5D533",
                  "borderBottomWidth": 2.34375,
                  "borderColor": "#D5D5D5",
                  "borderLeftWidth": 7.03125,
                  "borderRadius": 75,
                  "borderRightWidth": 2.34375,
                  "borderTopWidth": 7.03125,
                  "height": 117.1875,
                  "paddingHorizontal": 18.75,
                },
                undefined,
              ]
            }
          >
            <TextInput
              accessibilityLabel="confirm-password-input"
              autoCapitalize="none"
              keyboardType="default"
              multiline={false}
              onChangeText={[Function]}
              placeholder="Confirm password"
              placeholderTextColor="#A9A8A3"
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "color": "#353434",
                    "fontSize": 32.8125,
                    "height": "100%",
                    "paddingHorizontal": 43.9453125,
                    "paddingTop": 0,
                    "position": "absolute",
                    "width": "100%",
                  },
                  undefined,
                ]
              }
              testID="confirm-password-input"
              textContentType="password"
              value=""
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "height": 75,
                },
                undefined,
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "height": 75,
                },
                undefined,
              ]
            }
          />
          <View
            accessibilityLabel="sign-up-button"
            accessibilityState={
              Object {
                "disabled": true,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={false}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#F5BE59",
                "borderRadius": 75,
                "height": 107.8125,
                "justifyContent": "center",
                "opacity": 0.6,
              }
            }
            testID="sign-up-button"
          >
            <Text
              size="md"
              style={
                Array [
                  Object {
                    "color": "#353434",
                    "fontSize": 32.8125,
                    "fontWeight": "700",
                  },
                  Object {
                    "color": "#353434",
                  },
                ]
              }
              type="bold"
            >
              Sign Up
            </Text>
          </View>
          <View
            accessibilityLabel="already-have-account-button"
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#FAFAFA",
                "borderRadius": 75,
                "height": 107.8125,
                "justifyContent": "center",
                "opacity": 1,
              }
            }
            testID="already-have-account-button"
          >
            <Text
              size="md"
              style={
                Array [
                  Object {
                    "color": "#353434",
                    "fontSize": 32.8125,
                    "fontWeight": "700",
                  },
                  Object {
                    "color": "#353434",
                  },
                ]
              }
              type="bold"
            >
              Already have an account? Log in
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
